adr/0004-event-bus-ownership.md

```diff
@@ -0,0 +1,38 @@
+# 4. Own event bus exchanges and queues in services
+
+Date: 2018-05-17
+
+## Status
+
+Proposed
+
+## Context
+
+In a service-based architecture, infrastructure needs to be owned by a service so that it is provisioned and maintained as a first class citizen.
+
+The [event system](0001-events-system.md) based on RabbitMQ requires the setup of `exchanges` and `queues` that `bind` to them.
+
+Practical experience developing eLife lead to no mutual dependencies between services, with e.g. service B strictly being downstream from service A with no events feeding back into service A.
+
+Some services may require a generic binding to many other exchanges in the interest of observability, e.g. a dashboard aggregating events from many different services.
+
+## Decision
+
+Create an `exchange` for each data type, as part of a service publishing events.
```

Comment:
Any service can publish to the `exchange`? Thinking Search when starting/finishing indexing an article.

Reply:
Option 1, separate exchanges per-project:

- Article Store publishes to `articles`
- Downstream Sample publishes to `downstream-sample`
- Search published to `search`

Different event types require a standard correlation_id to be grouped together. Requires some assembly in the dashboard or anything else listening to multiple sources.

Option 2, rely on the original data type exchange:

- Article Store publishes to `articles` with type `article.published`
- Downstream Sample publishes to `articles` with type `article.crossref.started`
- Search published to `articles` with type `article.search.indexed`

Sightly easier to setup a downstream service, but it will receive and discard its own events. The dependency to the `articles` exchange is already there, so the surface area of a service like Article Store already includes its exchanges (can potentially be wrapped in an HTTP API rather than writing directly to an exchange from a service that doesn't own it). Requires configuration of the data types in the dashboard. Requires active ignoring of all unknown event types as a design decision (wouldn't do RabbitMQ magic with routing keys).